File: mcp-llm-server/src/mcp_llm_server/routes/usage.py
"""Usage-related routes."""

from datetime import date

from fastapi import APIRouter
from pydantic import BaseModel

from mcp_llm_server.config.settings import get_settings
from mcp_llm_server.infra.usage_repository import DailyUsage, get_usage_repository


router = APIRouter(prefix="/api/usage", tags=["Usage"])


class UsageResponse(BaseModel):
    """Token usage response"""

    input_tokens: int
    output_tokens: int
    total_tokens: int
    reasoning_tokens: int | None
    model: str | None = None


class DailyUsageResponse(BaseModel):
    """Daily usage response"""

    usage_date: str
    input_tokens: int
    output_tokens: int
    total_tokens: int
    reasoning_tokens: int
    request_count: int
    model: str | None = None


class UsageListResponse(BaseModel):
    """List of daily usages"""

    usages: list[DailyUsageResponse]
    total_input_tokens: int
    total_output_tokens: int
    total_tokens: int
    total_request_count: int
    model: str | None = None


def _resolve_usage_model(model: str | None = None) -> str:
    return model or get_settings().gemini.default_model


def _build_daily_usage_response(
    usage: DailyUsage | None, model: str | None = None
) -> DailyUsageResponse:
    resolved_model = _resolve_usage_model(model)
    if usage is None:
        return DailyUsageResponse(
            usage_date=str(date.today()),
            input_tokens=0,
            output_tokens=0,
            total_tokens=0,
            reasoning_tokens=0,
            request_count=0,
            model=resolved_model,
        )
    return DailyUsageResponse(
        usage_date=usage.usage_date.isoformat(),
        input_tokens=usage.input_tokens,
        output_tokens=usage.output_tokens,
        total_tokens=usage.total_tokens,
        reasoning_tokens=usage.reasoning_tokens,
        request_count=usage.request_count,
        model=resolved_model,
    )


def _build_usage_list_response(
    usages: list[DailyUsage], model: str | None = None
) -> UsageListResponse:
    resolved_model = _resolve_usage_model(model)
    usage_list = [
        _build_daily_usage_response(usage, resolved_model) for usage in usages
    ]
    return UsageListResponse(
        usages=usage_list,
        total_input_tokens=sum(u.input_tokens for u in usages),
        total_output_tokens=sum(u.output_tokens for u in usages),
        total_tokens=sum(u.total_tokens for u in usages),
        total_request_count=sum(u.request_count for u in usages),
        model=resolved_model,
    )


@router.get("/daily", response_model=DailyUsageResponse)
async def api_usage_daily() -> DailyUsageResponse:
    """Get today's token usage from DB"""
    repository = get_usage_repository()
    usage = await repository.get_daily_usage()

    model = get_settings().gemini.default_model
    return _build_daily_usage_response(usage, model=model)


@router.get("/recent", response_model=UsageListResponse)
async def api_usage_recent(days: int = 7) -> UsageListResponse:
    """Get recent N days token usage from DB"""
    repository = get_usage_repository()
    usages = await repository.get_recent_usage(days=days)

    model = get_settings().gemini.default_model
    return _build_usage_list_response(usages, model=model)


@router.get("/total", response_model=UsageResponse)
async def api_usage_total(days: int = 30) -> UsageResponse:
    """Get aggregated token usage for N days from DB"""
    repository = get_usage_repository()
    usage = await repository.get_total_usage(days=days)

    return UsageResponse(
        input_tokens=usage.input_tokens,
        output_tokens=usage.output_tokens,
        total_tokens=usage.total_tokens,
        reasoning_tokens=usage.reasoning_tokens,
    )

File: mcp-llm-server-go/internal/handler/usage.go
package handler

import (
	"log/slog"
	"net/http"
	"strconv"
	"time"

	"github.com/gin-gonic/gin"

	"github.com/park285/llm-kakao-bots/mcp-llm-server-go/internal/config"
	"github.com/park285/llm-kakao-bots/mcp-llm-server-go/internal/httperror"
	"github.com/park285/llm-kakao-bots/mcp-llm-server-go/internal/usage"
)

// DailyUsageResponse: 일자별 사용량 응답입니다.
type DailyUsageResponse struct {
	UsageDate       string `json:"usage_date"`
	InputTokens     int64  `json:"input_tokens"`
	OutputTokens    int64  `json:"output_tokens"`
	TotalTokens     int64  `json:"total_tokens"`
	ReasoningTokens int64  `json:"reasoning_tokens"`
	RequestCount    int64  `json:"request_count"`
	Model           string `json:"model"`
}

// UsageListResponse: 사용량 목록 응답입니다.
type UsageListResponse struct {
	Usages            []DailyUsageResponse `json:"usages"`
	TotalInputTokens  int64                `json:"total_input_tokens"`
	TotalOutputTokens int64                `json:"total_output_tokens"`
	TotalTokens       int64                `json:"total_tokens"`
	TotalRequestCount int64                `json:"total_request_count"`
	Model             string               `json:"model"`
}

// UsageHandler: 사용량 API 핸들러입니다.
type UsageHandler struct {
	cfg    *config.Config
	repo   *usage.Repository
	logger *slog.Logger
}

// NewUsageHandler: 사용량 핸들러를 생성합니다.
func NewUsageHandler(cfg *config.Config, repo *usage.Repository, logger *slog.Logger) *UsageHandler {
	return &UsageHandler{
		cfg:    cfg,
		repo:   repo,
		logger: logger,
	}
}

// RegisterRoutes: 사용량 라우트를 등록합니다.
func (h *UsageHandler) RegisterRoutes(router *gin.Engine) {
	group := router.Group("/api/usage")
	group.GET("/daily", h.handleDaily)
	group.GET("/recent", h.handleRecent)
	group.GET("/total", h.handleTotal)
}

func (h *UsageHandler) handleDaily(c *gin.Context) {
	usageRow, err := h.repo.GetDailyUsage(c.Request.Context(), time.Time{})
	if err != nil {
		h.logError(err)
		writeError(c, err)
		return
	}

	c.JSON(http.StatusOK, h.buildDailyResponse(usageRow))
}

func (h *UsageHandler) handleRecent(c *gin.Context) {
	days, ok := parseDays(c, 7)
	if !ok {
		return
	}

	usages, err := h.repo.GetRecentUsage(c.Request.Context(), days)
	if err != nil {
		h.logError(err)
		writeError(c, err)
		return
	}

	c.JSON(http.StatusOK, h.buildUsageListResponse(usages))
}

func (h *UsageHandler) handleTotal(c *gin.Context) {
	days, ok := parseDays(c, 30)
	if !ok {
		return
	}

	usageRow, err := h.repo.GetTotalUsage(c.Request.Context(), days)
	if err != nil {
		h.logError(err)
		writeError(c, err)
		return
	}

	c.JSON(http.StatusOK, UsageResponse{
		InputTokens:     usageRow.InputTokens,
		OutputTokens:    usageRow.OutputTokens,
		TotalTokens:     usageRow.TotalTokens(),
		ReasoningTokens: usageRow.ReasoningTokens,
		Model:           h.cfg.Gemini.DefaultModel,
	})
}

func (h *UsageHandler) buildDailyResponse(usageRow *usage.DailyUsage) DailyUsageResponse {
	model := h.cfg.Gemini.DefaultModel
	if usageRow == nil {
		return DailyUsageResponse{
			UsageDate:       time.Now().Format("2006-01-02"),
			InputTokens:     0,
			OutputTokens:    0,
			TotalTokens:     0,
			ReasoningTokens: 0,
			RequestCount:    0,
			Model:           model,
		}
	}

	return DailyUsageResponse{
		UsageDate:       usageRow.UsageDate.Format("2006-01-02"),
		InputTokens:     usageRow.InputTokens,
		OutputTokens:    usageRow.OutputTokens,
		TotalTokens:     usageRow.TotalTokens(),
		ReasoningTokens: usageRow.ReasoningTokens,
		RequestCount:    usageRow.RequestCount,
		Model:           model,
	}
}

func (h *UsageHandler) buildUsageListResponse(usages []usage.DailyUsage) UsageListResponse {
	model := h.cfg.Gemini.DefaultModel
	response := UsageListResponse{
		Usages:            make([]DailyUsageResponse, 0, len(usages)),
		TotalInputTokens:  0,
		TotalOutputTokens: 0,
		TotalTokens:       0,
		TotalRequestCount: 0,
		Model:             model,
	}

	for _, row := range usages {
		response.Usages = append(response.Usages, DailyUsageResponse{
			UsageDate:       row.UsageDate.Format("2006-01-02"),
			InputTokens:     row.InputTokens,
			OutputTokens:    row.OutputTokens,
			TotalTokens:     row.TotalTokens(),
			ReasoningTokens: row.ReasoningTokens,
			RequestCount:    row.RequestCount,
			Model:           model,
		})
		response.TotalInputTokens += row.InputTokens
		response.TotalOutputTokens += row.OutputTokens
		response.TotalTokens += row.TotalTokens()
		response.TotalRequestCount += row.RequestCount
	}

	return response
}

func parseDays(c *gin.Context, defaultDays int) (int, bool) {
	raw := c.Query("days")
	if raw == "" {
		return defaultDays, true
	}
	parsed, err := strconv.Atoi(raw)
	if err != nil || parsed <= 0 {
		writeError(c, httperror.NewInvalidInput("days must be a positive integer"))
		return 0, false
	}
	return parsed, true
}

func (h *UsageHandler) logError(err error) {
	if err == nil {
		return
	}
	h.logger.Warn("usage_request_failed", "err", err)
}
